### ISSUE TITLE
When some one call me with his actual number, how can I talk to him instead of reading a text to him via TwiML?

### ISSUE THREAD

**Issue by allenming123** [twilio/voice-quickstart-objc#103]
![8f61aac8-2c0d-4170-8b02-f7903a51ed40](https://user-images.githubusercontent.com/36840501/39515764-a0ad3d32-4e2d-11e8-850d-1053cf852180.jpg)

> Configuration of twilio number and TwiML App as above

Let's say I have a twilio number: +1 415 965 1234
Bob has his own number(not a twilio number): +86 12345678910

When Bob calls +1 415 965 1234, he gets the voice of "Congratulations! You have just made your first call! Good bye."  and the phone call is terminated.

I know twilio request the url in picture above
and this is my server's code
![691799b7-f19e-418a-90a2-5beb3c0eb963](https://user-images.githubusercontent.com/36840501/39516321-308dc132-4e2f-11e8-801c-f15d9f7be3ed.jpg)

I have 3 confusions, and I really hope you can help me out:

1. How can I talk to him instead of reading the text to him via TwiML? 

2. It seems when I call someone with twilio number or that twilio number gets an income call, it all comes to the makeCall function in the server code, how can I distinguish the outgoing call and income call?

3. where can I find more info about VoiceResponse data? I can't even find "to" value in voice_response.py, and I am newbie to python.
```
resp = VoiceResponse()
to = request.values.get("to")
```

**Comment by bobiechen-twilio** [comment]
Hi @allenming123 

Thanks for reaching out and these are fairly good questions/confusions SDK users usually encounter.

For the specific case of an incoming call to your Twilio number from a regular PSTN number, you might have noticed that the `to` value is not set and the `/makeCall` method just returns a TwiML with a `<Say>` verb.

1. To route this call to another person, you can simply remove the line that checks if the `to` value is not set and return a `<Dial>` verb to a number or to another registered client identity.

2. Actually both cases (PSTN to your Twilio number, or Voice SDK `TwilioVoice.call()` method) are consider **incoming** to your Twilio number, and the logic of actions is exactly defined in the TwiML app URL. In this case, the `/makeCall` endpoint will determine what to do based on the `to` value. This value might not be set if the incoming call is from a non-programmable end (PSTN number). On your mobile app/SDK side, you can define extra parameters you need when calling `TwilioVoice.call()` and give it to the `params:` dictionary, then extract it on your server side by calling `request.value.get(PARAM_YOU_DEFINED)`.

3. The `to` value is extracted from the HTTP request, instead of the `VoiceResponse` object, but if your are interested, the [`VoiceResponse` object is defined in the `twilio-python` library](https://github.com/twilio/twilio-python/blob/93cf463f914f55c4c4bd1c259b834953dd81609d/twilio/twiml/voice_response.py#L16).

Another way to separate the incoming traffic is to add another Twilio number and use different TwiML app URLs for each number.

Hope this helps.

-bobie

**Comment by allenming123** [comment]
Hi @bchen-twilio 

About question 1 and 3

I know the to value is extracted from the HTTP request, but in my situation:

Bob calls my twilio number from his iPhone's Phone APP -> my  twilio number get an incoming call -> twilio request the URL
In this work flow, I wanna know what data would twilio pass to my server's web application, and where to find the document for that data. Now I know the 'to' value is empty, the 'from' value is empty.
```
from1 = request.values.get("from")
print from1    # I got a None
```
I know to use  **Dial** verb to talk to Bob, but without the 'from' value(Bob's PSTN number), I can't call him back in my web application.

OR maybe twilio number just don't support this case?

About question 2

If I can guarantee whenever I make an outgoing call from my VoIP app, the 'to' value is not empty. Can I do this to distinguish an outgoing call from incoming call? (I need ur experiences, plz pointing out the cases if I miss any. Thanks! )

![acf20ff9-48fe-4767-a9c6-ffafa97af670](https://user-images.githubusercontent.com/36840501/39561079-e859b202-4ed5-11e8-9f2f-e711309655bf.jpg)

**Comment by bobiechen-twilio** [comment]
Hi @allenming123 

Not sure why your last reply is not shown here.
1 & 3. I believe the default parameters that Twilio passes to the Voice URL are `To` and `From`. Yes, capitalized.
2. I think with the clarification of the default incoming URL parameters above, you can choose more specialized parameter key-value pairs that will be sent from your mobile application, so that it's easier to differentiate whether the URL request is from a regular PSTN number or from your Voice mobile app.

Hope this makes it clearer for you
-bobie

**Comment by allenming123** [comment]
Hi @bchen-twilio 

Bob use his PSTN number(+8612345678910) call my twilio number(+14159651234)
-> Twilio request the url (endpoint: makeCall)
-> In makeCall function, I can get value from the request: From: +8612345678910, To: +14159651234

```
From = request.values.get("From")
To = request.values.get("To")
```

Then I make this response:
`resp.dial(callerId=From).number(To)`, this will make a Infinite loop
or  I make this response:
`resp.dial(callerId=To).number(From)`, this just doesn't make any sense

The problem is I just wan to answer the call from +8612345678910, what should I do?

Later I find this response can make me receive the phone call, but when I tap the answer button on my iphone with callKit, it just hangup.
`resp.dial(callerId=From).client('alice')`

Is there a better way to answer the call from +8612345678910? 
I think it is a basic function that when twilio number receive a PSTN call, I can just answer the call and talk to the caller.

**Comment by bobiechen-twilio** [comment]
Hi @allenming123 

Glad that we are one step further!
Doing `resp.dial(callerId=From).number(To)` will make another call to `+14159651234`, which is your Twilio number and the same Voice URL will be hit again, thus the infinite loop.

My question is, who should receive the call and answer it when Bob `+8612345678910` is calling `+14159651234`? basically your should use a different value than `To` when doing `resp.dial(callerId=From).number(__THE_NUMBER_TO_CALL__)`.

To summarize:
- if you want to receive and answer the call from Bob on a PSTN phone, use `resp.dial(callerId=From).number(__THE_NUMBER_TO_CALL__)`.
- if you want to receive the call on a Programmable Voice mobile client, use `resp.dial(callerId=From).client(__THE_IDENTITY_TO_CALL__)`.

About the disconnect right away problem on your CallKit client, can you provide us the SDK log or were you getting any failure callbacks like `call:didFailWithError:`?

Thanks
-bobie

**Comment by bobiechen-twilio** [comment]
HI @allenming123 

I am closing this ticket and assuming that your app is capable of making client to client/number calls without problems. Please feel free to reach out to us if you run into any problems or questions in the future.

Thanks.
-bobie
